File: 백준/Gold/2636. 치즈/치즈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer token;
    static StringBuilder sb = new StringBuilder();

    static int N,M,chzCnt;
    static int[][] graph;
    static boolean[][] visited;
    static Deque<int[]> deq;

    static int[] deltaX = {0,0,1,-1};
    static int[] deltaY = {1,-1,0,0};

    public static void main(String[] args) throws IOException {
        token = new StringTokenizer(input.readLine());
        N = Integer.parseInt(token.nextToken());
        M = Integer.parseInt(token.nextToken());
        chzCnt = 0;
        graph = new int[N][M];
        for (int i = 0; i < N; i++) {
            token = new StringTokenizer(input.readLine());
            for (int j = 0; j < M; j++) {
                int num = Integer.parseInt(token.nextToken());
                graph[i][j] = num;
                if (num==1){
                    chzCnt++;
                }
            }
        }
        int time = 0;
        int lastChzCnt = Integer.MAX_VALUE;
        while (chzCnt!=0){
            visited = new boolean[N][M];
            deq = new ArrayDeque<>();
            deq.addLast(new int[]{0,0});
            visited[0][0] = true;
            while (!deq.isEmpty()){
                int[] cur = deq.pollFirst();
                int x = cur[0];
                int y = cur[1];
                for (int i=0; i<4; i++){
                    int dx = x+deltaX[i];
                    int dy = y+deltaY[i];
                    if (isValid(dx,dy) && !visited[dy][dx]){
                        if (graph[dy][dx]==0){
                            deq.addLast(new int[]{dx,dy});
                            visited[dy][dx] = true;
                        }else{
                            graph[dy][dx] ++;
                        }
                    }
                }
            }
            lastChzCnt = chzCnt;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    if (graph[i][j]>=2){
                        graph[i][j] = 0;
                        chzCnt--;
                    }
                }
            }
            time++;
        }
        sb.append(time).append("\n").append(lastChzCnt);
        System.out.println(sb);
    }
    public static boolean isValid(int x, int y){
        return x>=0 && x<M && y>=0 && y<N;
    }
}

File: 백준/Gold/2636. 치즈/치즈.py
'''
아이디어:
패딩 추가 후, 치즈가 녹을 때까지 반복
'''
import sys
from collections import deque

def is_valid(x,y): # 탐색 가능 범위
    return 0<=x<M and 0<=y<N

delta_x = [0,0,-1,1]
delta_y = [1,-1,0,0]

N,M = map(int, sys.stdin.readline().split())
graph = []
graph.append([0]*(M+2)) # 패딩 추가
for _ in range(N):
    graph.append([0]+list(map(int, sys.stdin.readline().split()))+[0])
graph.append([0]*(M+2))

N += 2; M += 2 # 패딩 추가 범위
cnt_chz = 0
for i in range(1,N-1):
    for j in range(1,M-1):
        if graph[i][j]:
            cnt_chz += 1

time = 0
last_chz_cnt = 0 # 마지막 치즈 갯수
while cnt_chz:
    time += 1
    melt_idxs = [] # 나중에 한 번에 지울 idx list
    visited = []
    for _ in range(N):
        visited.append([False]*M)
    deq = deque()
    deq.append([0,0]) # 항상 공기와 접촉
    while deq:
        cur = deq.popleft()
        if graph[cur[1]][cur[0]] == 1:
            melt_idxs.append([cur[0],cur[1]])
        for k in range(4):
            dx = cur[0] + delta_x[k]
            dy = cur[1] + delta_y[k]
            if is_valid(dx,dy) and not visited[dy][dx] and graph[cur[1]][cur[0]] == 0: # 이전 것이 공기가 통할 때
                deq.append([dx,dy])
                visited[dy][dx] = True
    for idx in melt_idxs: # 치즈 녹이기
        graph[idx[1]][idx[0]] = 0
    cnt_chz -= len(melt_idxs)
    last_chz_cnt = len(melt_idxs)

print(time)
print(last_chz_cnt)